janusgraph-core/src/main/java/org/janusgraph/graphdb/tinkerpop/optimize/strategy/AdjacentVertexHasUniquePropertyOptimizerStrategy.java

```diff
@@ -0,0 +1,141 @@
+import org.apache.tinkerpop.gremlin.process.traversal.step.filter.FilterStep;
+import org.apache.tinkerpop.gremlin.process.traversal.step.filter.HasStep;
+import org.apache.tinkerpop.gremlin.process.traversal.step.filter.NoneStep;
+import org.apache.tinkerpop.gremlin.process.traversal.step.map.GraphStep;
+import org.apache.tinkerpop.gremlin.process.traversal.step.map.IdStep;
+import org.apache.tinkerpop.gremlin.process.traversal.step.util.HasContainer;
+import org.apache.tinkerpop.gremlin.process.traversal.util.TraversalHelper;
+import org.apache.tinkerpop.gremlin.structure.Edge;
+import org.apache.tinkerpop.gremlin.structure.Graph;
+import org.apache.tinkerpop.gremlin.structure.Vertex;
+import org.janusgraph.core.Cardinality;
+import org.janusgraph.core.JanusGraphElement;
+import org.janusgraph.graphdb.database.StandardJanusGraph;
+import org.janusgraph.graphdb.query.JanusGraphPredicateUtils;
+import org.janusgraph.graphdb.query.QueryUtil;
+import org.janusgraph.graphdb.query.condition.*;
+import org.janusgraph.graphdb.query.index.IndexSelectionUtil;
+import org.janusgraph.graphdb.tinkerpop.optimize.JanusGraphTraversalUtil;
+import org.janusgraph.graphdb.transaction.StandardJanusGraphTx;
+import org.janusgraph.graphdb.types.CompositeIndexType;
+import org.janusgraph.graphdb.types.IndexType;
+
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import static org.janusgraph.graphdb.types.system.ImplicitKey.ADJACENT_ID;
+import static org.janusgraph.graphdb.types.system.ImplicitKey.ID;
+
+/**
+ * @author Florian Grieskamp (Florian.Grieskamp@gdata.de)
+ */
+public class AdjacentVertexHasUniquePropertyOptimizerStrategy
+    extends AdjacentVertexOptimizerStrategy<HasStep<?>> {
+
+    private static final AdjacentVertexHasUniquePropertyOptimizerStrategy INSTANCE =
+        new AdjacentVertexHasUniquePropertyOptimizerStrategy();
+
+    private AdjacentVertexHasUniquePropertyOptimizerStrategy() {}
+
+    public static AdjacentVertexHasUniquePropertyOptimizerStrategy instance() {
+        return INSTANCE;
+    }
+
+    @Override
+    public Set<Class<? extends ProviderOptimizationStrategy>> applyPost() {
+        Set<Class<? extends ProviderOptimizationStrategy>> postStrategies = new HashSet<>();
+        postStrategies.add(JanusGraphStepStrategy.class);
+        postStrategies.add(JanusGraphLocalQueryOptimizerStrategy.class);
+        return postStrategies;
+    }
+
+    @Override
+    public void apply(final Traversal.Admin<?, ?> traversal) {
+        if (TraversalHelper.onGraphComputer(traversal) || !traversal.getGraph().isPresent()) {
+            return;
+        }
+
+        Graph graph = traversal.getGraph().get();
+        StandardJanusGraph janusGraph = graph instanceof StandardJanusGraph
+            ? (StandardJanusGraph) graph
+            : ((StandardJanusGraphTx) graph).getGraph();
+
+        if (!janusGraph.getConfiguration().optimizerBackendAccess()) {
```

Comment:
I think this check could be placed on a more high level. Instead of always triggering `AdjacentVertexHasUniquePropertyOptimizerStrategy.java` for each query we can just skip adding this optimization strategy when this check is disabled in JanusGraph. When `janusGraph.getConfiguration().optimizerBackendAccess()` is `false` we don't even need to add this optimization to our life-cycle. Unless I miss something

Reply:
I.e. I see that `TraversalStrategies` are registered in a static module in `StandardJanusGraph` but I think we could refactor that later. I guess we need just a separate `refactor` issue for that

Reply:
Created #2338